strawberry/schema/subscribe.py

```diff
@@ -0,0 +1,154 @@
+                    extensions_runner,
+                    process_errors,
+                )
+            else:
+                aiterator = agen_or_result.__aiter__()
+                running = True
+                while running:
+                    # reset extensions results for each iteration
+                    execution_context.extensions_results = {}
+                    async with extensions_runner.executing():
+                        try:
+                            origin_result: Union[
+                                ExecutionResult, OriginalExecutionResult
+                            ] = await aiterator.__anext__()
+
+                        except StopAsyncIteration:
+                            break
+                        except Exception as exc:
+                            # graphql-core doesn't handle exceptions raised in the async generator.
+                            origin_result = ExecutionResult(
+                                data=None, errors=[_coerce_error(exc)]
+                            )
+                            running = False
+                    # we could have yielded in the except block above.
+                    # but this way we make sure `get_result` hook is called deterministically after
+                    # `on_execute` hook is done.
+                    yield await _handle_execution_result(
+                        execution_context,
+                        origin_result,
+                        extensions_runner,
+                        process_errors,
+                    )
+        # catch exceptions raised in `on_execute` hook.
+        except Exception as exc:
+            origin_result = OriginalExecutionResult(
+                data=None, errors=[_coerce_error(exc)]
+            )
+            yield await _handle_execution_result(
+                execution_context,
+                origin_result,
+                extensions_runner,
+                process_errors,
+            )
+
+
+async def subscribe(
+    schema: GraphQLSchema,
+    execution_context: ExecutionContext,
+    extensions_runner: SchemaExtensionsRunner,
+    process_errors: ProcessErrors,
+    middleware_manager: MiddlewareManager,
+) -> SubscriptionResult:
+    asyncgen = _subscribe(
+        schema,
+        execution_context,
+        extensions_runner,
+        process_errors,
+        middleware_manager,
+    )
+    # GrapQL-core might return an initial error result instead of an async iterator.
+    # This happens when "there was an immediate error" i.e resolver is not an async iterator.
+    # To overcome this while maintaining the extension contexts we do this trick.
+    first = await asyncgen.__anext__()
+    if isinstance(first, ExecutionResultError):
+        # close the async generator (calling `.aclose()` on it won't close the context managers)
+        with contextlib.suppress(StopAsyncIteration):
+            await asyncgen.__anext__()
+        return first
```

Comment:
Why `.aclose()` will not close it?

Also, maybe use [aclosing](https://docs.python.org/3/library/contextlib.html#contextlib.aclosing) (it is only available to python 3.10+ but we can backport it) to make sure we always close it no matter what?

```python
# Copied from cpython's code for contextlib.aclosing
class aclosing(Generic[_I], contextlib.AbstractAsyncContextManager):
    def __init__(self, thing: _I):
        super().__init__()
        self.thing = thing

    async def __aenter__(self) -> _I:
        return self.thing

    async def __aexit__(self, *exc_info):
        await self.thing.aclose()
```

If we really need the extra `__anext__` we can pass an extra `await_next_on_exit: bool = False` to it and do that before the `.aclose` itself

update: I also checked some docs here and `.aclose` should make the context close, unless it was using `yield` without a `try: finally` clause. But in this case it is using a class with `__exit__`/`__aexit__`

Reply:
ah this was because of try finally thing. I think it is solved by this https://github.com/nrbnlulu/strawberry/blob/61154dcfab222051b56a77d13c49a40fe203fda8/strawberry/extensions/context.py#L206 though